File: Prefix Suffix String - GFG/prefix-suffix-string.java
//{ Driver Code Starts
//Initial Template for Java
import java.util.*;
import java.io.*;

public class code1 
{
    public static void main(String args[])throws IOException
    {
        BufferedReader in=new BufferedReader(new InputStreamReader(System.in));
        PrintWriter out=new PrintWriter(System.out);
        int t=Integer.parseInt(in.readLine());
        while(t-->0){
            String s1[]=in.readLine().trim().split(" ");
            String s2[]=in.readLine().trim().split(" ");

            Solution ob=new Solution();
            int ans=ob.prefixSuffixString(s1,s2);
            out.println(ans);

        }
        out.close();
    }
}
// } Driver Code Ends


class Trie {
    HashMap<Character, Trie> children;
    public Trie() {
        children = new HashMap<Character, Trie>();
    }
}

class Solution {   
    Trie root;
    public Solution() {
        root = new Trie();
    }
    
    public void insert(String word) {
        Trie cur = root;
        for(int i = 0; i < word.length(); i++) {
            char c = word.charAt(i);
            if(!cur.children.containsKey(c)) cur.children.put(c, new Trie());
            cur = cur.children.get(c);
        }
    }
    
    public boolean search(String word) {
        Trie cur = root;
        for(int i = 0; i < word.length(); i++) {
            char c = word.charAt(i);
            if(!cur.children.containsKey(c)) return false;
            cur = cur.children.get(c);
        }
        return true;
    }
    
    public int prefixSuffixString(String s1[],String s2[]) {
        for(String s : s1) {
            insert(s);
            String t = "";
            for(int i = s.length() - 1; i >= 0; i--) t += s.charAt(i);
            insert(t);
        }
        int ans = 0;
        for(String s : s2) {
            String t = "";
            for(int i = s.length() - 1; i >= 0; i--) t += s.charAt(i);
            if(search(s) || search(t)) ans++;
        }
        return ans;
    }
}

File: Prefix Suffix String - GFG/prefix-suffix-string.cpp
//{ Driver Code Starts
//Initial Template for C++

#include<bits/stdc++.h>
using namespace std;


// } Driver Code Ends
//User function Template for C++

class Solution {
public:
    struct TrieNode {
    	TrieNode *child[26];
    	TrieNode() {
    		for(int i = 0; i < 26; i++) child[i] = NULL;
    	}
    };
    
    TrieNode* insert(TrieNode* root, string &key) {
    	TrieNode* cur = root;
    	int len = key.size();
    	for(int i = 0; i < len; i++) {
    		if(cur -> child[key[i] - 'a'] == NULL) {
    			TrieNode* newNode = new TrieNode();
    			cur -> child[key[i] - 'a'] = newNode;
    		}
    		cur = cur -> child[key[i]-'a'];
    	}
    	return root;
    }
    
    bool search(TrieNode* root, string &key) {
    	TrieNode* cur = root;
    	int len = key.size();
    	for(int i = 0; i < len; i++) {
    		if(cur -> child[key[i] - 'a'] == NULL)
    			return false;
    		cur = cur -> child[key[i] - 'a'];
    	}
    	return true;
    }
    
    int prefixSuffixString(vector<string> &s1,vector<string> s2){
        // Code here
        TrieNode* root = new TrieNode();
        for(string s : s1) {
            root = insert(root, s);
            string t = s;
            reverse(t.begin(), t.end());
            root = insert(root, t);
        }
        int ans = 0;
        for(string s : s2) {
            if(search(root, s)) ans++;
            else {
                string t = s;
                reverse(t.begin(), t.end());
                if(search(root, t)) ans++;
            }
        }
        return ans;
    }
};

//{ Driver Code Starts.


int main(){
    int t;
    cin>>t;
    for(int i=0;i<t;i++){
        string str1,str2,newline;
        if(i==0)
        getline(cin,newline);
        getline(cin,str1);
        getline(cin,str2);
        // getline(cin,newline);
        // cout<<str1<<endl<<str2<<endl;
        vector<string> s1,s2;
        stringstream ss1(str1);
        stringstream ss2(str2);
        string str;
        while(ss1>>str){
            s1.push_back(str);
        }
        while(ss2>>str){
            s2.push_back(str);
        }
        Solution ob;
        cout<<ob.prefixSuffixString(s1, s2)<<endl;
    }
}
// } Driver Code Ends